File: blackjack/python-pascal/partitions.pas
library partitions;

uses
  ctypes;

type
   cint32 =  longint;
   stack  = array[0..9] of cint32;

function partitions(cards: stack; subtotal: cint32): cint32;
var
  i, total : integer;
begin
   partitions := 0;
   // Hit
   for i := 0 to 9 do
      begin
	 if (cards[i]>0) then
	 begin
	    total := subtotal+i+1;
	    if (total < 21) then
	    begin
	       // Stand
	       partitions += 1;
	       // Hit again
	       cards[i] -= 1;
	       partitions += partitions(cards, total);
	       cards[i] += 1;
	    end
	    else
	       if (total=21) then
		  // Stand; hit again is an automatic bust
		  partitions += 1;
	 end;
      end;
end;

exports
  partitions;

end.

File: blackjack/c-c/partitions.c
int partitions(int cards[10], int subtotal)
{
  int m=0;
  int total;
  // Hit
  for (int i = 0; i < 10; i++) {
    if (cards[i]>0) {
      total = subtotal+i+1;
      if (total < 21) {
	cards[i] -= 1;
	// Stand
	m += 1;
	// Hit again
	m += partitions(cards, total);
	cards[i] += 1;
      } else if (total==21) {
	// Stand; hit again is an automatic bust
	m += 1;
      }
    }
  }
  return m;
}
